link-grammar/parse/count.c

```diff
@@ -630,6 +775,15 @@ static Count_bin do_count(
 						}
 					}
 				}
+				if (need_re_w_count && !rcount_found && (0 != hist_total(&l_bnr)))
```

Comment:
It seems to me that you do not actually need to test `&& (0 != hist_total(&l_bnr))` here, unless you are trying to somehow prevent `rcount_found` being set to true.

Reply:
Oh. Maybe this is the point-2. that you are making above? Hmmm.

Reply:
We need to find out if the RHS that matches `re` may **ever** have any count that can be multiplied by an LHS count (if exists). Here we know that there is no count due to Rmatch (since ` !rcount_found`.  But maybe there is a count due to  `l_bnr`.  `If `(0 == hist_total(&l_bnr)` then we know there is no such count, so this calculation is skipped. If `(0 != hist_total(&l_bnr)` it may be that the count exists (and the cached `l_bnr` will be used by the `CACHE_COUNT` macro) or it is `NO_COUNT` (and the macro will call `do_count()` to find it in the hard way).

Note that only `rcount_found` is set in case there is an `l_bnr` count, and the multiplication/accumulation is not done. This is because currently this check is performed only when `le == NULL`, and in that case there cannot be a total count ("subtle" we already said).
``` c
    544                                 else if (re_w_count == -1 && le == NULL) need_re_w_count = true;
```
Also, since we only need to know in this case and others if `rcount_found`, a further optimization could be done if a special argument of `do_count()` to tell it to return immediately on positive count (without continuing to calculate the count). But such optimizations start to be unmanageable...).

Applying this PR to the current master (with an added argument to `form_match_list()` to disable the just added "lc optimization" when calculating the cached `le_w_count`) yeilds **37% speedup of the fix-long** run. The total runtime for me is 75 seconds, about half of it is due to the `And yet [...]` sentence (with still has a panic on the standard timeout because it takes 32 seconds).
I also validated that the linkages are exactly the same as without this PR (checked with `-test=len-trailing-hash:0` in order to generate suffix_id's for all the sentences).

But despite of the big speedup, I still propose not to use this idea yet. I have more ideas to test, and I hope to get a deeper understanding when testing them.

Reply:
OK.